tests/frontend/test_tdmprogram.py

```diff
@@ -756,3 +758,68 @@ def test_shots_passed(self):
         results = eng.run(prog, shots=2)
         assert results.samples.shape[0] == 2
         assert prog.run_options["shots"] == 5
+
+
+class TestTDMValidation:
+    """Test the validation of TDMProgram against the device specs"""
+    @pytest.fixture(scope="class")
+    def device(self):
+        target = "TD2"
+        tm = 4
+        layout = f"""
+            name template_tdm
+            version 1.0
+            target {target} (shots=1)
+            type tdm (temporal_modes=2)
+            float array p0[1, {tm}] =
+                {{rs_array}}
+            float array p1[1, {tm}] =
+                {{r_array}}
+            float array p2[1, {tm}] =
+                {{bs_array}}
+            float array p3[1, {tm}] =
+                {{m_array}}
+            Sgate(p0) | 1
+            Rgate(p1) | 0
+            BSgate(p2, 0) | (0, 1)
+            MeasureHomodyne(p3) | 0
+        """
+        device_spec = {
+            "layout": inspect.cleandoc(layout),
+            "modes": {"concurrent": 2, "spatial": 1, "temporal_max": 100},
+            "compiler": [target],
+            "gate_parameters": {
+                "p0": [-1],
+                "p1": [1],
+                "p2": [2],
+                "p3": [3],
+            },
+        }
+        return DeviceSpec("TD2", device_spec, connection=None)
+    
+    @staticmethod
+    def compile_test_program(device, args=(-1, 1, 2, 3)):
+        """Compiles a test program with the given gate arguments."""
+        alpha = [args[1]]
+        beta = [args[2]]
+        gamma = [args[3]]
+        prog = tdmprogram.TDMProgram(N=2)
+        with prog.context(alpha, beta, gamma) as (p, q):
+            ops.Sgate(args[0]) | q[1]  # Note that the Sgate has a second parameter that is non-zero
+            ops.Rgate(p[0]) | q[0]
+            ops.BSgate(p[1]) | (q[0], q[1])
+            ops.MeasureHomodyne(p[2]) | q[0]
+        prog.compile(device=device, compiler=device.compiler)
+
+    def test_validation_correct_args(self, device):
+        """Test that no error is raised when the tdm circuit explicit parameters within the allowed ranges"""
+        self.compile_test_program(device, args=(-1, 1, 2, 3))
+
+    @pytest.mark.parametrize("incorrect_index", list(range(4)))
+    def test_validation_incorrect_args(self, device, incorrect_index):
+        """Test the correct error is raised when the tdm circuit explicit parameters are not within the allowed ranges"""
+        args = [-1, 1, 2, 3]
+        args[incorrect_index] = -999
+        with pytest.raises(CircuitError, match="Parameter has value '-999' while its valid range is "):
+            self.compile_test_program(device, args=args)
+    
```

Comment:
There seems to be a slight issue here with the ending newline. Perhaps just remove the final tab and make sure there's only a single empty line at the end? 🤔